client/executor/wasmtime/src/runtime.rs

```diff
@@ -87,282 +64,108 @@ pub fn create_instance(
-
-	// Enable/disable producing of debug info.
-	context.set_debug_info(false);
-
-	// Instantiate and link the env module.
-	let global_exports = context.get_global_exports();
-	let compiler = new_compiler(compilation_strategy)?;
-	let env_module = instantiate_env_module(global_exports, compiler, host_functions)?;
-	context.name_instance("env".to_owned(), env_module);
-
-	// Compile the wasm module.
-	let module = context.compile_module(&code)
-		.map_err(|e| WasmError::Other(format!("module compile error: {}", e)))?;
-
-	Ok((module, context))
-}
-
 /// Call a function inside a precompiled Wasm module.
 fn call_method(
-	context: &mut Context,
-	module: &mut CompiledModule,
+	module: &Module,
+	externs: &[Extern],
+	state_holder: &StateHolder,
 	method: &str,
 	data: &[u8],
 	heap_pages: u32,
 ) -> Result<Vec<u8>> {
-	// Old exports get clobbered in `InstanceHandle::new` if we don't explicitly remove them first.
-	//
-	// The global exports mechanism is temporary in Wasmtime and expected to be removed.
-	// https://github.com/CraneStation/wasmtime/issues/332
-	clear_globals(&mut *context.get_global_exports().borrow_mut());
-
-	let mut instance = module.instantiate()
-		.map_err(|e| Error::Other(e.to_string()))?;
-
-	// Ideally there would be a way to set the heap pages during instantiation rather than
-	// growing the memory after the fact. Currently this may require an additional mmap and copy.
-	// However, the wasmtime API doesn't support modifying the size of memory on instantiation
-	// at this time.
-	grow_memory(&mut instance, heap_pages)?;
-
-	// Initialize the function executor state.
-	let heap_base = get_heap_base(&instance)?;
-	let executor_state = FunctionExecutorState::new(heap_base);
-	reset_env_state_and_take_trap(context, Some(executor_state))?;
+	let instance_wrapper = unsafe {
+		let instance = Instance::new(module, externs)
+			.map_err(|e| WasmError::Other(format!("cannot instantiate: {}", e)))?;
 
-	// Write the input data into guest memory.
-	let (data_ptr, data_len) = inject_input_data(context, &mut instance, data)?;
-	let args = [RuntimeValue::I32(u32::from(data_ptr) as i32), RuntimeValue::I32(data_len as i32)];
-
-	// Invoke the function in the runtime.
-	let outcome = context
-		.invoke(&mut instance, method, &args[..])
-		.map_err(|e| Error::Other(format!("error calling runtime: {}", e)))?;
-	let trap_error = reset_env_state_and_take_trap(context, None)?;
-	let (output_ptr, output_len) = match outcome {
-		ActionOutcome::Returned { values } => match values.as_slice() {
-			[RuntimeValue::I64(retval)] => unpack_ptr_and_len(*retval as u64),
-			_ => return Err(Error::InvalidReturn),
-		}
-		ActionOutcome::Trapped { message } => return Err(trap_error.unwrap_or_else(
-			|| format!("Wasm execution trapped: {}", message).into()
-		)),
+		// This is safe since the requirement that `InstanceWrapper` own the instance exclusively
+		// is held.
+		InstanceWrapper::new(instance)?
```

Comment:
A better `InstanceWrapper` API would avoid the need for this to be unsafe.

Reply:
Please note that the decision using here `unsafe` (that spans two modules) was completely deliberate. The other option - constructing the instance within the `InstanceWrapper::new` - thus avoiding unsafe here of course was also considered. However, I settled on the present approach to make the invariant more visible (if not screaming). In the other approach there won't be a suitable place to put an unsafe block.